### ISSUE TITLE
Rethink Pagination

### ISSUE THREAD

**Issue by rmosolgo** [rmosolgo/graphql-ruby#1359]
Currently you have a few options to paginate with GraphQL-Ruby: 

- Don't paginate at all, return list types everywhere (:-1:)
- Roll your own pagination system (:+1:, but tricky, and we could probably help each other out here)
- `GraphQL::Relay::Connection` & friends (😬 ) 

In my opinion, `GraphQL::Relay::Connection` has some problems: 

- Important parts of the system are hidden from view (namely, the handling and application of `first`/`after`/`last`/`before`)
- The automagic object-to-wrapper mapping based on class hides pagination, so: 
   - It's easy to add a regression where something that _should_ be an AR::Relation starts being an Array, but GraphQL-Ruby doesn't miss a beat. If you want to paginate something as a Relation, it would be nice to have a stricter mapping, to avoid accidentally loading the whole relation into memory. 
  - Sometimes, a class-to-wrapper mapping isn't enough: 
    - Rails 3 does weird things, I think AR::Relation _is_ an Array, or something 
    - If you want to apply a specialization in some cases (when the returned object is of the same class), you don't have a good option. You have to introduce a proxy class, and a custom wrapper, etc. 
- The system is not easy to customize, customization requires some code duplication or un-Ruby-ish factory functions. 

So, I would like to reconsider our pagination system with a few goals: 

- Make the application object -> pagination wrapper mapping _explicit_, in the resolver 
- Don't hide any arguments from the resolver; pass everything in there. 
- Use a class-based system for extending and customizing pagination, for improved hackability. (Combined with an explicit object -> wrapper mapping, this could support more case-by-case customization.)

I would like to preserve some things from the current system: 

- Backwards compat, for sure 
- Cursor-based, Relay-style pagination (although I'd be open to adding some other options for non-Relay users?) 

So, does anyone else have thoughts about how pagination could be improved in GraphQL-Ruby?

**Comment by eapache** [comment]
So we've completely rolled our own pagination internally, and I'm not sure how well the patterns will apply to the gem generally but I'll share the shape for context at least.

Internally, we have these things called `Filter`s which are little classes which can be appended to fields. They may define arguments, and they get an opportunity to manipulate the result of resolving the field before it's actually returned. The canonical example internally is `FirstFilter` whose implementation (in its entirety) is:
```ruby
  class FirstFilter < Filter
    argument :first, :integer, description: "Truncate the array result to this size"

    def apply_to(field)
      raise "`#{field.name}` must return an array in order to use `FirstFilter`" unless valid_field_type?(field.type)
      super
    end

    def call(results, args, _)
      args[:first] ? results.first(args[:first]) : results
    end

    private

    def valid_field_type?(field_type)
      field_type.unwrap_non_null.is_a?(ListType)
    end
  end
```

It's pretty straight-forward; when the field is constructed, all of the arguments from its filters are added to the definition, and `apply_to` is called on the filters for any custom logic. When the field is resolved, the filters are chained on the end with `call`.

Internally, all our pagination logic is implemented as filters. We have a `BidirectionalConnection` filter which takes care of defining and validating first/last/before/after, defining common methods for cursor serialization, etc. and (roughly, I'm glossing over some details) a `SqlConnection < BidirectionalConnection` filter which adds logic for doing relative pagination through AR associations.

Then, defining a connection field looks something like:
```ruby
field :name, Connection.for(Type), filters: [SqlConnection], resolve: ->(obj, _, _) { obj.association }
```
The filter takes care of adding the arguments, and of applying them to the AR association object and transforming the result into the right shape in terms of edges and cursors.

The nice thing is that it's pretty swappable; we also have (again roughly) an `ElasticSearchConnection < BidirectionalConnection`.

Honestly filters themselves were designed pretty generically, but 90% of their actual use has been for pagination. Their only other really common use for us has been defining height/width transforms on image fields.

**Comment by eapache** [comment]
@swalkinshaw pointed out to me that our `Filter` is kind of like the gem's `Function` but I see a few differences (though these may be stylistic more than anything given you can subclass `Function`s for specific fields):
- `Function` defines the entire field, where-as `Filter` is an "add-on", i.e. `Filter` appends to the existing `resolve` instead of replacing it
- `Filters` can be stacked (we've used this exactly once AFAIK, for a connection of images that needed both pagination and image transform args)

**Comment by rmosolgo** [comment]
Wow, that's a really cool system, thanks for sharing 😎 

> can be stacked 

That's a really cool take on composition! I will chew on this for a bit.

**Comment by rmosolgo** [comment]
I took a try at this here: https://github.com/rmosolgo/graphql-ruby/pull/1411

This is a much more approachable API than instrumentation, too, since there's no `proc` wrapping 😆

**Comment by theorygeek** [comment]
This seems really nice. I like the syntax too, I imagine that we would use it like:

```ruby
def self.preload(association)
  PreloadFilter.for(association)
end

def self.auth_scope(scope)
  AuthorizationFilter.for(scope)
end

field :foo, FooType, filters: [preload(:bar), auth_scope(:sensitive_info)]
```

Although, I am betting that this is what people are going to try to do:
```ruby
field :foo, FooType, filters: [PreloadFilter.new(:bar), AuthorizeFilter.new(:sensitive_info)]
```

Notice the difference, the `Filter` is instantiated once and then used on more than one field. It feels a little more intuitive than the factory-ish pattern of `.for(something).new(field)`.

**Comment by bgentry** [comment]
#1014 is related. That's the code I'm using to power pagination in my app using sequel and stable, bidirectional cursors.

Definitely agree with your statement of problems with the current model.

**Comment by rmosolgo** [comment]
Closed by #2143
